model/Login.php

```diff
@@ -198,4 +198,95 @@ public function changePassword($username, $new_password, $old_password): bool
 
         return true;
     }
+
+    /**
+     * @param string $username
+     * @param string $new_password
+     * @param string $old_password
+     *
+     * All passwords need to be plain text; they'll be hashed appropriately
+     * as per the configuration in config.inc.php
+     *
+     * @return boolean true on success; false on failure
+     * @throws \Exception if invalid user, or db update fails.
+     */
+    public function addAppPassword($username, $password, $app_desc, $app_pass): bool
+    {
+        list(/*NULL*/, $domain) = explode('@', $username);
+
+        if (!$app_pass) {
+            throw new \Exception(Config::Lang('pAppPassAdd_pass_empty_error'));
+        }
+        if (!$app_desc) {
+            throw new \Exception(Config::Lang('pException_desc_empty_error'));
+        }
+
+        if (!$this->login($username, $password)) {
+            throw new \Exception(Config::Lang('pPassword_password_current_text_error'));
+        }
+
+        $app_pass = pacrypt($app_pass);
+
+        $fields    = "username, description, password_hash";
+        $values    = "\"$username\", \"$app_desc\", \"$app_pass\"";
+
+/* maybe we want this
+        if (Config::bool('password_expiration')) {
+            $domain = $this->getUserDomain($username);
+            if (!is_null($domain)) {
+                $password_expiration_value = (int)get_password_expiration_value($domain);
+                $set['password_expiry'] = date('Y-m-d H:i', strtotime("+$password_expiration_value day"));
+            }
+        }
+*/
+        $result = db_execute("REPLACE INTO mailbox_app_password($fields) VALUES($values)");
```

Comment:
i think this would be vulnerable ot SQL injection - you need to use placeholders instead....
e.g. `db_execute('REPLACE INTO mailbox_app_password(username, description, password_hash) VALUES (:u, :d, :p)", ['u' => ... , 'd' => ...., 'p' => ..... )");`

Reply:
(I'm not sure if postgresql, mysql and sqlite all support the replace into syntax either).